### PR TITLE
Extract errors to a library

### PR DESCRIPTION
**Pull request by ThunderDeliverer**
Errors were extractred to a library smart contract. The contract option was picked over the library simply because deployment in such manner should conserve gas. The errors library smart contracts was then imported in the smart contracts and the inheritance tree respected, so that each instance only gets it inherited once.

### CODE REVIEW

**Comment by steven2308** [comment]
This is much cleaner than I expected, really nice. I just removed the contract definition since i's not really necessary. May I merge it?

**Comment by ThunderDeliverer** [comment]
Please don't merge until I push the commit containing the error documentation, I'll be done in the evening and we can merge then (if I don't do it now, I'll have to adapt the documentation plan 😅 )

**Comment by ThunderDeliverer** [comment]
This is now ready for review 👌

**Comment by ThunderDeliverer** [comment]
@steven2308 I have no objection to your unrelated commits 🙌 🚀

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:20]: Let's merge this into ERC721ApprovalToCurrentOwner.

Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol:20]: Could this not be emitted if the caller was approved for all and tried to call approval on specific token for self?

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:91]: Let's remove " resource" and add this:
/// @dev when a token is nested, only the direct owner (NFT parent) can mange it, approved addresses are not allowed in that case to ensure consistency.

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:92]: /// Attempting to compose a resource without a base associated

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:96]: This is actually the contract owner, shall we rename it?
It guards operations as locking the contract, transfer ownership and withdrawing the value from the contract.

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:114]: Typo equipped

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:116]: /// Attempting to equip an item into a slot that the target resource does not implement

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:118]: /// Attempting to equip a child into a slot and parent that the child does not support.

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:122]: /// Attempting to determine the resource with the top priority on a token with no resources added.

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [inline contracts/RMRK/library/RMRKErrors.sol:124]: typo: empty

**Review by steven2308 — CHANGES_REQUESTED** [review]
In some cases the name of the error might be misleading, if that's the case let's discuss better options.

**Comment by stevyhacker** [comment]
This looks much cleaner and I like that the errors have a single place for documentation too!

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [reply contracts/RMRK/library/RMRKErrors.sol:20]: What you describe is the case on MultiResource so I guess we need to keep it.
`if (owner == operator) revert ERC721ApproveToCaller();` 
On Nesting and all inheriting implementations, we have an extra level on approvals which is always set to the sender, so there we could use the one I suggested. But it might look inconsistent :thinking: 
`if (_msgSender() == operator) revert ERC721ApproveToCaller();`

Both checks are on `setApprovalForAll`

Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol:20]: I'd leave these two separate for now and we can address them if we find a good and consistent solution. At least we have a centralised errors file where we can do it :D

File: contracts/RMRK/library/RMRKErrors.sol
Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol:31]: Sure, makes it more verbose

File: contracts/RMRK/library/RMRKErrors.sol
Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol:40]: Done :D

File: contracts/RMRK/library/RMRKErrors.sol
Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol:43]: How about this?

File: contracts/RMRK/library/RMRKErrors.sol
Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol:96]: Hm, I'll fix the error description for now, but we can create a separate task to rename owner of the contract into something like `Issuer`, `Creator`, `Maintainer`, ...

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [reply contracts/RMRK/library/RMRKErrors.sol:43]: What do you mean?
I'm just pointing at a missing "s" :stuck_out_tongue:

Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol]: Hahaha, I was replying to comments in the VS code while doing the changes, so the response might make more sense now, that I pushed them 😂

File: contracts/RMRK/library/RMRKErrors.sol
Comment (steven2308) [reply contracts/RMRK/library/RMRKErrors.sol:43]: @ThunderDeliverer  only this one is missing.

Comment (ThunderDeliverer) [reply contracts/RMRK/library/RMRKErrors.sol:43]:
```suggestion
/// Attempting to add a resource entry with `Part`s, without setting the `Base` address
```

**Review by steven2308 — APPROVED** [review]
Great change in general!

**Pull request merged**
